### ISSUE TITLE
Question about feature selection

### ISSUE THREAD

**Issue by randombishop** [dmlc/xgboost#434]
I am experimenting feature selection strategies, and wondering if the order of features can have an effect on xgboost performance ?
In other words, will xgboost do the same with feature set {A,B,C,D,E,F,G} as with feature set {E,F,G,A,B,C,D} ?
(I have around 200 features, many of which are very sparse)

**Comment by pommedeterresautee** [comment]
Feature order should not have any effect if you are not using sub-sampling on columns (all data points are checked to find best split in each tree). 
And if you do, results may be slightly different (because of randomness in sub-sampling).

Check slides from @tqchen (https://github.com/dmlc/xgboost/blob/master/doc/index.md) for more info.

Moreover, if you are doing feature selection you may be interested in this conversation:
https://github.com/dmlc/xgboost/issues/419

In particular the idea implemented in GBM to add noise and see how useful are these fake features.

Let us know your results if you try it.

Kind regards,
Michaël

**Comment by randombishop** [comment]
Thank you very much Michaël, you are right, I had colsample_bytree=0.6.
It's also true that the results are slightly different, there's a 0.00001 difference in my cross validation.
The problem is that this difference turns into 0.001 in the public leaderboard of caterpillar kaggle competition !!! It's still small but worth many positions :)
Thanks also for the noise-feature tip, i will definitely use it for my feature selection and keep you updated.

**Comment by randombishop** [comment]
Follow up about noise-feature tip : it works really well not only for feature selection, but also for parameters tuning. 
With my previous params, noise-feature was ranked n°1 in feature importance.
I changed the params until it went down to a meaningfull ranking, then selected the more important features, and it looks like i have a better model now.
Thank you very much for your advice !

**Comment by NoskovLab** [comment]
Nice idea!

**Comment by soerendip** [comment]
Sometimes, a noise feature can rank very high. But removing the features less often used, leads to a drop in cross-validated and  test set performance. So, I think the measure is actually not good. Some features are used less often, but still have valuable information.

**Comment by davidfjr3** [comment]
What did you mean more clearly? Did you say that variable ranking does not
mean anything?

Another thought: To evaluate the importance of the variables depends on
which measure you're using it for, right? importance of measures such as
"often" are not very good? The best measure would be some weight"
measure within the model trees?

Perhaps the "weight" is the Gini index, right?

As you can see, these concepts are not very clear to me.

Thank you.

2016-07-01 14:12 GMT-03:00 Sören <notifications@github.com>:

**Comment by khotilov** [comment]
The approach to parameter tuning is this case would be the same as when you want to reduce overfitting in training data. The more overfitting, the more algorithm  would cling to irrelevant features.

Also much would depend on your data. If the data or labels are rather noisy, it would be especially important not to rely on a single run with injected noise features, but to do many runs to find a threshold. Also, instead of generating completely artificial random features,  I would recommend creating noise features as random shuffles of some typical features in your data.

**Comment by soerendip** [comment]
The term "importance" is a bit misleading, because the f-score simply calculates how often a feature is used in a model. [Link](http://stackoverflow.com/questions/34218245/how-is-the-feature-score-in-the-xgboost-package-calculated) 

The problem is that 'importance' implies that a feature that scores higher contributes more to a good prediction than a feature that scores lower, but that is not necessarily the case. In fact, a random feature can score higher than features that are rare, but that contains information. F-score should not be called 'feature importance', because it is fishy. Instead, just call it f-score. Or 'feature-split-count' or something along the lines.

There is also the discrepancy between 'importance to fit the model' and 'importance to make predictions'. Actually, importance always needs to be used in conjunction with something 'importance for something'. The best would be to avoid the term 'importance' and name the scores according to what they actually measure. Whether or not it is important, leave it to the user.

**Comment by igorfmr** [comment]
there's a way to use xgboost to select features similar to %incMse in random forest ??
